darshan-util/pydarshan/darshan/tests/test_plot_dxt_heatmap.py

```diff
@@ -460,13 +455,13 @@ def test_plot_heatmap(filepath, mod, ops):
         # horizontal bar graph does not exist
         assert jgrid.ax_marg_x.has_data()
         assert jgrid.ax_joint.has_data()
-        if "ior_hdf5_example.darshan" in filepath:
-            # verify the horizontal bar graph contains data since there
-            # are multiple ranks for this case
-            assert jgrid.ax_marg_y.has_data()
-        else:
-            # verify the horizontal bar graph does not contain data
+        if "dxt.darshan" in filepath:
+            # verify the horizontal bar graph does not contain data since there
+            # are is only 1 rank for this case
```

Comment:
May need some proofreading on the comments here I think.

I think I see what is happening with the test changes here, though it is a bit hard to read because file names have changed and there is no parametrization association, so two files can be associated with 1 value that got flipped, which does make it harder for me to read.

More concerning is this change locally for multi-core testing:

- `pydarshan-devel`: `pytest --pyargs darshan -n 6`
  - `298 passed in 92.99s (0:01:32)`
- this branch: `pytest --pyargs darshan -n 6`
  - I gave up after 5.5 minutes

Single core is a bit more reasonable. Nonetheless, it makes me a bit hesitant to move forward here just yet. I may have to at least check if using a genuine sparse data structure and taking advantage of `seaborn.heatmap` feature of `Cells with missing values are automatically masked` would do the same thing faster/with less sustained memory usage.

Reply:
Okay, I'll go through the comments again and push in more revisions. 

I've been having issues running `pytest --pyargs darshan -n 6` as well, but it seems to work fine with 4 cores, or if I run without using `xdist`:
- `pytest --pyargs darshan -n 4`: `297 passed in 102.48s (0:01:42)`
- `pytest --pyargs darshan`: `297 passed in 134.63s (0:02:14)`

Also I found that after running `pytest --pyargs darshan -n 6 --cache-clear` once I can now run `pytest --pyargs darshan -n 6` fine.

Reply:
Right, a single core is probably using way less memory; if you run on N cores and generate N large reports (incl. DXT figs) simultaneously, there may be some issues, so that's where I have pause.

Reply:
I tried converting the heatmap dataframe to `SparseDtype` before the reindex:
```diff
diff --git a/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py b/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py
index a917242..e79bed0 100644
--- a/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py
+++ b/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py
@@ -381,5 +381,6 @@ def get_heatmap_df(agg_df: pd.DataFrame, xbins: int, nprocs: int) -> pd.DataFram
     cats = cats.mul(agg_df["length"], axis=0)
     cats.index = agg_df["rank"]
     hmap_df = cats.groupby("rank").sum()
+    hmap_df = hmap_df.astype(pd.SparseDtype("float", fill_value=0.0))
     hmap_df = hmap_df.reindex(index=range(nprocs), fill_value=0.0)
     return hmap_df
```

Printing out the memory usage with `df.info(memory_usage="deep")`, this change appears to reduce the memory usage:
```
FILE: snyder_acme.exe_id1253318_9-27-24239-1515303144625770178_2.darshan
==============================
mod:  DXT_POSIX
hmap_df (original) memory usage: 12.6 MB
hmap_df (SparseDtype) memory usage: 274.4 KB
hmap_df (SparseDtype, reindexed) memory usage: 210.5 KB
==============================
mod:  DXT_MPIIO
hmap_df (original) memory usage: 48.7 KB
hmap_df (SparseDtype) memory usage: 28.9 KB
hmap_df (SparseDtype, reindexed) memory usage: 28.8 KB
```

The tests still pass with this change (`pytest --pyargs darshan -n 6`: `298 passed in 102.51s (0:01:42)`), and the heatmaps look fine. 

The performance isn't better, of course, but maybe that's tolerable if the memory is of greater concern:

`asv continuous pydarshan-devel nawtrey_issue_575_update_ymax -b "time_plot_heatmap_builtin_logs"`
```
       before           after         ratio
     [536cdce1]       [cba5828d]
     <pydarshan-devel>       <nawtrey_issue_575_update_ymax>
+        906±20ms       1.38±0.06s     1.52  dxt_heatmap.PlotDXTHeatMapSmall.time_plot_heatmap_builtin_logs('tests/input/sample-dxt-simple.darshan', 1000)
+      1.01±0.01s       1.42±0.05s     1.41  dxt_heatmap.PlotDXTHeatMapSmall.time_plot_heatmap_builtin_logs('examples/example-logs/dxt.darshan', 1000)
+      1.01±0.02s       1.41±0.03s     1.41  dxt_heatmap.PlotDXTHeatMapSmall.time_plot_heatmap_builtin_logs('examples/example-logs/ior_hdf5_example.darshan', 1000)
+         262±4ms         335±10ms     1.28  dxt_heatmap.PlotDXTHeatMapSmall.time_plot_heatmap_builtin_logs('tests/input/sample-dxt-simple.darshan', 10)
+         363±4ms         459±10ms     1.26  dxt_heatmap.PlotDXTHeatMapSmall.time_plot_heatmap_builtin_logs('tests/input/sample-dxt-simple.darshan', 100)
+        402±10ms         460±20ms     1.14  dxt_heatmap.PlotDXTHeatMapSmall.time_plot_heatmap_builtin_logs('examples/example-logs/ior_hdf5_example.darshan', 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

Reply:
Not sure yet--if memory were chosen over performance then `asv` (peak) memory benchmarks may also be in order: https://asv.readthedocs.io/en/stable/benchmarks.html#benchmark-types 

Deeper question may be if sparsity should start with data structures that are lower down, near the CFFI layer. The code is so slow that `asv` benchmarks for larger logs would be painful, on top of still not having access to the logs repo in `asv`...

Many of those sparse slowdowns are still not as bad as the original version of the PR though. Adding the `snyder..` file to the benchmarks with a custom timeout to allow the benchmarks to run for enough replicates may shed some light on it, but would be painful.

Reply:
Of course, sparse data structures take less memory, though it is not immediately obvious to me what happens when `seaborn` encounters the sparse data structure and has to plot it--maybe it does something clever, though it is certainly quite possible that it produces the dense zeros eventually.

Reply:
> Not sure yet--if memory were chosen over performance then asv (peak) memory benchmarks may also be in order: https://asv.readthedocs.io/en/stable/benchmarks.html#benchmark-types

Yeah, if we want to go this route then that's probably a requirement. If the sparse representation lowers a relative max instead of the global peak then it's probably not worth the trouble. 

> Of course, sparse data structures take less memory, though it is not immediately obvious to me what happens when `seaborn` encounters the sparse data structure and has to plot it--maybe it does something clever, though it is certainly quite possible that it produces the dense zeros eventually.

I tried tracing the path through `seaborn.matrix.heatmap()`, and it appears that it effectively sends `hmap_df.values` directly into `matplotlib.pyplot.pcolormesh()`, which then runs it through `np.asanyarray()`, and finally through [this function](https://github.com/matplotlib/matplotlib/blob/710fce3df95e22701bd68bf6af2c8adbc9d67a79/lib/matplotlib/cbook/__init__.py#L700-L711) (array is `x`) where it uses `np.array()`. I'm not sure exactly what numpy does with sparse arrays when it encounters them, but my understanding is it doesn't recognize that they're sparse. So it's not clear what is happening yet, but I'll try to walk through those steps and see what happens. It might be as simple as retrieving the plot data from one of the plotting classes, but I'll have to see.

Reply:
asarray() will guarantee dense representation yeah

Reply:
I pushed in a fix for the typo mentioned here.